File: Week 01/id_636/LeetCode_1_636.go
package id_636

/*
	从列表中筛选出两个数字，使得相加等于 Target，
	如果能找到则返回数字的下标 ，找不到则返回空。
	两种办法：
		1：暴力破解，时间复杂度 O(n^2)
		2：使用 map 方法，可以记录已经遍历的数字，一次遍历即可
*/

// 省略暴力方法的代码，直接用第二种
func twoSum(nums []int, target int) []int {
	// 略过判断数组长度的部分， 呃，还是写上吧
	if len(nums) < 2 {
		return []int{}
	}

	// 记录已经存在
	exist := make(map[int]int)

	// 循环遍历数组
	for i := 0; i < len(nums); i++ {
		// index & value
		value := nums[i]
		index := i

		// check
		possibleNum := target - value
		if key, ok := exist[possibleNum]; ok {
			return []int{key, index}
		}
		exist[value] = index
	}
	return []int{}
}

File: Week 01/id_636/LeetCode_1_636.py
class Solution:
    def twoSum(self, nums: list, target: int) -> list:
        # check exist
        exist = dict()
        res = []
        # for one loops
        for index in range(len(nums)):
            # store current value
            value = nums[index]
            # check two numbers whether exist
            if (target - value) in exist:
                if index == exist.get(target-value):
                    continue
                return [exist.get(target-value), index]
            exist.update({
                value: index
            })
        # not exist
        return []

if __name__ == '__main__':
    '''
    For test.
    仅返回头一个遇到的结果。
    '''
    s = Solution()
    nums = [0, 1, 2, 3, 5, 8]
    target = 6
    print("res: ", s.twoSum(nums, target))
